From c12a12dd9cb881dbec5380f585549d88c5a446f2 Mon Sep 17 00:00:00 2001
From: Xulin Sun <xulin.sun@windriver.com>
Date: Wed, 21 Nov 2018 16:08:53 +0800
Subject: [PATCH] vfio: vfio_fsl_mc: fix a kmemleak caused in
 vfio_fsl_mc_ioctl

memdup_user() allocates memory to give vfio_fsl_mc_set_irqs_ioctl(),
after that the allocated memory should be freed.

The kmemleak is reported:

unreferenced object 0xffff8023247ec900 (size 64):
  comm "lcore-slave-2", pid 2684, jiffies 4294928274 (age 2699.224s)
  hex dump (first 32 bytes):
    3a 00 00 00 23 80 ff ff 00 00 00 00 23 80 ff ff :...#.......#...
    10 c9 7e 24 23 80 ff ff 10 c9 7e 24 23 80 ff ff ..~$#.....~$#...
  backtrace:
    [<ffff000008231030>] create_object+0xf8/0x278
    [<ffff000008aefd84>] kmemleak_alloc+0x74/0xa0
    [<ffff000008223614>] __kmalloc_track_caller+0x1ac/0x348
    [<ffff0000081df9d8>] memdup_user+0x38/0xf0
    [<ffff0000087c4f8c>] vfio_fsl_mc_ioctl+0x124/0x3d8
    [<ffff0000087b9794>] vfio_device_fops_unl_ioctl+0x44/0x70
    [<ffff00000824f508>] do_vfs_ioctl+0xb0/0x848
    [<ffff00000824fd2c>] SyS_ioctl+0x8c/0xa8
    [<ffff000008083b18>] __sys_trace_return+0x0/0x4
    [<ffffffffffffffff>] 0xffffffffffffffff

Signed-off-by: Xulin Sun <xulin.sun@windriver.com>
---
 drivers/vfio/fsl-mc/vfio_fsl_mc.c |    1 +
 1 file changed, 1 insertion(+)

diff --git a/drivers/vfio/fsl-mc/vfio_fsl_mc.c b/drivers/vfio/fsl-mc/vfio_fsl_mc.c
index aa682c9..5d6f6fb 100644
--- a/drivers/vfio/fsl-mc/vfio_fsl_mc.c
+++ b/drivers/vfio/fsl-mc/vfio_fsl_mc.c
@@ -273,6 +273,7 @@ static long vfio_fsl_mc_ioctl(void *device_data, unsigned int cmd,
 		ret = vfio_fsl_mc_set_irqs_ioctl(vdev, hdr.flags,
 						 hdr.index, hdr.start,
 						 hdr.count, data);
+		kfree(data);
 		return ret;
 	}
 	case VFIO_DEVICE_RESET:
-- 
1.7.9.5

